lib/external-secret.js

```diff
@@ -1,53 +1,64 @@
 'use strict'
 
+const JSONStream = require('json-stream')
+
 /**
- * Block asynchronous flow.
- * @param {number} milliseconds - Number of milliseconds to block flow operation.
- * @returns {Promise} Promise object representing block flow operation.
+ * Creates an FIFO queue which you can put to and take from.
+ * If theres nothing to take it will wait with resolving until
+ * something is put to the queue.
+ * @returns {Object} Queue instance with put and take methods
  */
-function sleep ({ milliseconds }) {
-  return new Promise(resolve => setTimeout(resolve, milliseconds))
+function createEventQueue () {
+  const queuedEvents = []
+  const waitingResolvers = []
+
+  return {
+    take: () => queuedEvents.length > 0
+      ? Promise.resolve(queuedEvents.shift())
+      : new Promise(resolve => waitingResolvers.push(resolve)),
+    put: (msg) => waitingResolvers.length > 0
+      ? waitingResolvers.shift()(msg)
+      : queuedEvents.push(msg)
+  }
 }
 
 /**
  * Get a stream of external secret events. This implementation uses
- * polling/batching and converts it to a stream of events.
+ * watch and yields as a stream of events.
  * @param {Object} kubeClient - Client for interacting with kubernetes cluster.
  * @param {Object} customResourceManifest - Custom resource manifest.
- * @param {number} intervalMilliseconds - Interval time in milliseconds for polling external secrets.
- * @param {Object} logger - Logger for logging stuff.
  * @returns {Object} An async generator that yields externalsecret events.
  */
 function getExternalSecretEvents ({
   kubeClient,
   customResourceManifest,
-  intervalMilliseconds,
   logger
 }) {
   return (async function * () {
+    const eventQueue = createEventQueue()
+
+    const stream = kubeClient
+      .apis[customResourceManifest.spec.group]
+      .v1.watch[customResourceManifest.spec.names.plural]
+      .getStream()
+
+    const jsonStream = new JSONStream()
+    stream.pipe(jsonStream)
+
+    jsonStream.on('data', eventQueue.put)
+
+    jsonStream.on('error', (err) => {
+      logger.warn(err, 'Got error on stream')
+      eventQueue.put({ type: 'DELETE_ALL' })
+    })
+
+    jsonStream.on('end', () => {
+      logger.warn('Watch stream ended')
```

Comment:
This could potentially make the logs noisy. The kube-apiserver disconnects streams periodically (see min-request-timeout here https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/)